doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,164 @@
+
+The ``cabal list-bin`` command is a convenient way to find the path to the built executable, which
+is needed to run the application with the right RTS flags as in the last command:
+``/path/to/my-app +RTS -pj -RTS``.
+
+Finally, when our application run finishes,
+a profiling JSON report (due to option ``-pj``) is written to a ``<app-name>.prof`` file,
+i.e. ``my-app.prof``, in the current directory.
+You can load this JSON report ``my-app.prof`` into a visualizer
+and look for performance bottlenecks. One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is
+`Speedscope <https://speedscope.app>`__,
+which runs in the browser and can open this JSON file directly.
+See the
+`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
+on how to optimize your code based on the profiling results afterwards.
+
+Different report formats can be generated by using different RTS flags. Some useful ones are:
+
+- ``-p`` for a GHC's own
+  `standard report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#cost-centres-and-cost-centre-stacks>`__
+  ``<app-name>.prof``, which can be visualized with `profiteur <https://github.com/jaspervdj/profiteur>`__
+  or `ghcprofview <https://github.com/portnov/ghcprofview-hs>`__.
+- ``-pj`` for a
+  `JSON report <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#json-profile-format>`__
+  ``<app-name>.prof``, which can be visualized with `Speedscope <https://speedscope.app>`__.
+- ``-l-p`` for a binary
+  `"eventlog" report <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#rts-eventlog>`__
+  ``<app-name>.eventlog``, which contains a lot more details and can show you resource usage over time, and can
+  be converted to JSON with `hs-speedscope <https://github.com/mpickering/hs-speedscope>`__
+  and visualized with `Speedscope <https://speedscope.app>`__.
+
+.. note::
+
+    The report generated from the ``cabal configure``, ``cabal build`` etc. commands can
+    also be generated with a single ``cabal run`` command, but no settings are persisted
+    to a file:
+
+    .. code-block:: console
+
+          $ cabal run --enable-profiling --profiling-detail=late my-app -- +RTS -pj -RTS
+          <program builds, runs and finishes>
+
+
+You can also pass a lot more
+`profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
+to GHC via the ``ghc-prof-options`` field in your ``my-app.cabal`` file.
+Add the following, recommended profiling options to your ``my-app.cabal`` file
+to control where performance measuring code is inserted into your application:
+
+::
+
+    executable my-app
+      ...
+      ghc-prof-options:
+        -fprof-auto
+        -fno-prof-count-entries
+        -fprof-auto-calls
+        ...
```

Comment:
Once again I ask to not put anything profile-related in the cabal file --- that's not what cabal files are for. These options can be added to a project(.local) file. Do you need help with figuring out how?

Reply:
Done. The only way i could find to pass such flags was "program options" (all in all a very confusing api). If that's not the intended name, please correct me.

Reply:
Looks good! I think it may be possible to do it per `package` stanza. But I haven't checked. 

I agree it's not super clean. But we should strive to maintain the distinction between project-level options and package descriptions. Even if they are technically not orthogonal and the same could be achieved with either means.